fdbserver/RestoreMaster.actor.cpp

```diff
@@ -0,0 +1,1339 @@
+		return Void();
+	}
+
+	printf("[INFO] Node:%s mutationLogPrefix:%s (hex value:%s)\n", self->describeNode().c_str(), mutationLogPrefix.toString().c_str(), getHexString(mutationLogPrefix).c_str());
+
+	// Determine the key range each applier is responsible for
+	int numLoaders = self->loadersInterf.size();
+	int numAppliers = self->appliersInterf.size();
+	ASSERT( numLoaders > 0 );
+	ASSERT( numAppliers > 0 );
+
+	state int loadingSizeMB = 0; //numLoaders * 1000; //NOTE: We want to load the entire file in the first version, so we want to make this as large as possible
+	int64_t sampleSizeMB = 0; //loadingSizeMB / 100; // Will be overwritten. The sampleSizeMB will be calculated based on the batch size
+
+	state double startTime = now();
+	state double startTimeBeforeSampling = now();
+	
+	wait( sampleWorkload(self, request, restoreConfig, sampleSizeMB) );
+	wait( delay(1.0) );
+
+	printf("[Progress] distributeWorkloadPerVersionBatch sampling time:%.2f seconds\n", now() - startTime);
+	state double startTimeAfterSampling = now();
+
+	// Notify each applier about the key range it is responsible for, and notify appliers to be ready to receive data
+	startTime = now();
+	wait( assignKeyRangeToAppliers(self, cx) );
+	wait( delay(1.0) );
+	printf("[Progress] distributeWorkloadPerVersionBatch assignKeyRangeToAppliers time:%.2f seconds\n", now() - startTime);
+
+	startTime = now();
+	wait( notifyAppliersKeyRangeToLoader(self, cx) );
+	wait( delay(1.0) );
+	printf("[Progress] distributeWorkloadPerVersionBatch notifyAppliersKeyRangeToLoader time:%.2f seconds\n", now() - startTime);
+
+	// Determine which backup data block (filename, offset, and length) each loader is responsible for and
+	// Notify the loader about the data block and send the cmd to the loader to start loading the data
+	// Wait for the ack from loader and repeats
+
+	// Prepare the file's loading status
+	for (int i = 0; i < self->files.size(); ++i) {
+		self->files[i].cursor = 0;
+	}
+
+	// Send loading cmd to available loaders whenever loaders become available
+	// NOTE: We must split the workload in the correct boundary:
+	// For range file, it's the block boundary;
+	// For log file, it is the version boundary.
+	// This is because
+	// (1) The set of mutations at a version may be encoded in multiple KV pairs in log files.
+	// We need to concatenate the related KVs to a big KV before we can parse the value into a vector of mutations at that version
+	// (2) The backuped KV are arranged in blocks in range file.
+	// For simplicity, we distribute at the granularity of files for now.
+
+	state int loadSizeB = loadingSizeMB * 1024 * 1024;
+	state int loadingCmdIndex = 0;
+
+	state int checkpointCurFileIndex = 0;
+	state long checkpointCurOffset = 0; 
+
+	startTime = now();
+	// We should load log file before we do range file
+	state RestoreCommandEnum phaseType = RestoreCommandEnum::Assign_Loader_Log_File;
+	state std::vector<Future<RestoreCommonReply>> cmdReplies;
+	loop {
```

Comment:
I'd probably suggest moving this whole `loop{}` into a `sendVersionBatchesToLoaders()` ?